src/Store.ts

```diff
@@ -0,0 +1,582 @@
+		if (this.Model && this.Model.createSubclass) {
+			// we need a distinct model for each store, so we can
+			// save the reference back to this store on it.
+			// we always create a new model to be safe.
+			this.Model = this.Model.createSubclass([]).extend({
+				// give a reference back to the store for saving, etc.
+				_store: this
+			});
+		}
+
+		// the object the store can use for holding any local data or events
+		this.storage = new Evented();
+		var store = this;
+		if (this.autoEmitEvents) {
+			// emit events when modification operations are called
+			this.autoEmitHandles.push(aspect.after(this, 'add', <any>this._emitUpdateEvent('add')));
+			this.autoEmitHandles.push(aspect.after(this, 'put', <any>this._emitUpdateEvent('update')));
+			this.autoEmitHandles.push(aspect.after(this, 'remove', function (result, args) {
+				const emit = function (result: Promise<any> | any) {
+					store._emit('delete', {
+						id: args[0],
+						type: undefined
+					});
+
+					return result;
+				};
+
+				if (result && result.then) {
+					result.then(emit);
+					return result;
+				}
+
+				return emit(result);
+			}));
+		}
+	}
+
+	private _emit(type: string, event: {type: string, cancelable?: boolean}): void | boolean {
+		event = event || {
+			type: undefined
+		};
+		event.type = type;
+		try {
+			return this.storage.emit(event);
+		} finally {
+			// Return the initial value of event.cancelable because a listener error makes it impossible
+			// to know whether the event was actually canceled
+			return event['cancelable'];
+		}
+	}
+
+	protected _emitUpdateEvent(type: string) {
+		return function (result: Promise<any> | any, args: {beforeId: string | number}[]) {
+			var self = this;
+			const emit = function (result: Promise<any> | any) {
+				var event: { [ index: string ]: any } = {
+					target: result
+				};
+				var	options = args[ 1 ] || <any>{};
+				if ('beforeId' in options) {
+					event['beforeId'] = options.beforeId;
+				}
+				self._emit(type, event);
+				return result;
+			};
+
+			if (result && result.then) {
+				result.then(emit);
+				return result;
+			}
+			else {
+				return emit(result);
+			}
+		};
+	}
+
+	protected _createSubCollection(kwArgs: {}) {
+		let newCollection = <{ [ index: string ]: any }> (<any>Object).setPrototypeOf({}, this.constructor);
+
+		for (let i in this) {
+			if (this._includePropertyInSubCollection(i, newCollection)) {
+				newCollection[i] = this[i];
+			}
+		}
+
+		return lang.mixin(newCollection, kwArgs);
+	}
+
+	protected _getQuerierFactory(type: string) {
+		var uppercaseType = type[0].toUpperCase() + type.substr(1);
+		return this['_create' + uppercaseType + 'Querier'];
+	}
+
+	protected _includePropertyInSubCollection(name: string, subCollection: { [ index: string ]: any }) {
+		return !(name in subCollection) || subCollection[name] !== this[name];
+	}
+
+	protected _restore(object: { [ index: string ]: any }, mutateAllowed: boolean) {
```

Comment:
Is this relying on dojo-core yet?  If so, its interfaces.d.ts provides `Hash` which could shorten this to `Hash<any>`.